dashboard/app/models/levels/bubble_choice.rb

```diff
@@ -192,6 +192,18 @@ def get_sublevel_for_progress(student, script)
     return best_result_sublevel(student, script)
   end
 
+  # Use pre-loaded feedback and progress data to get the sublevel ID using the
+  # same criteria as in get_sublevel_for_progress
+  def get_sublevel_for_progress_optimized(teacher_feedbacks:, user_levels:)
+    keep_working_level_id = teacher_feedbacks.find do |feedback|
+      feedback.review_state == TeacherFeedback::REVIEW_STATES.keepWorking
+    end&.level_id
+
```

Comment:
nit: Do we want to return early so we don't need to compute `max_progress_level_id` if we have `keep_working_level_id`?

Reply:
Sure, good idea.